### DISCUSSION TITLE
How do I prevent `Microsoft.Web.WebView2` from generating a directory containing cache data, or how do I move it?

### DISCUSSION THREAD

**Discussion by ero-qt** [MicrosoftEdge/WebView2Feedback#4470] [Q&A]
I have noticed that as soon as my C# application launches anything to do with `Microsoft.Web.WebView2`, a directory containing lots of cache data is generated next to my application. I do not want this, especially because the directory format is very confusing to users.

Is there a way to prevent this directory from being generated? Perhaps I can have it be generated in a more appropriate folder like `%LocalAppData%`?

**Comment by RichieHindle** [comment]
The [`CoreWebView2CreationProperties.UserDataFolder`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.winforms.corewebview2creationproperties.userdatafolder?view=webview2-dotnet-1.0.2903.40) property lets you control the location of the user data directory.

**Comment by blalond** [comment]
When creating:

```
tempUserDataFolder = Path.Combine(Path.GetTempPath(), "BrowserTemp", Guid.NewGuid().ToString());
Directory.CreateDirectory(tempUserDataFolder);

      CoreWebView2Environment env = await CoreWebView2Environment.CreateAsync(
                browserExecutableFolder: null,
                userDataFolder: tempUserDataFolder,
                options: options);
      await webView.EnsureCoreWebView2Async(env);
```

When closing:
```
private void WebView1_FormClosing(object? sender, FormClosingEventArgs e)
    {
        // Attempt to remove temporary user data folder if it exists
        if (!string.IsNullOrEmpty(tempUserDataFolder) && Directory.Exists(tempUserDataFolder))
        {
            try
            {
                var webViewProcessId = Convert.ToInt32(webView?.CoreWebView2.BrowserProcessId);
                var webViewProcess = System.Diagnostics.Process.GetProcessById(webViewProcessId);

                this.webView?.CoreWebView2.Profile.ClearBrowsingDataAsync();
                try { webView?.Dispose(); } catch { }

                // reference: https://stackoverflow.com/questions/70763018/how-to-delete-the-webview2-userdatafolder-in-c-sharp-wpf-application-browserpro
                // wait for process to exit with 3 second timeout
                webViewProcess.WaitForExit(3000);

                Directory.Delete(tempUserDataFolder, true);
            }
            catch
            {
                // best-effort: ignore failures to delete (files may be locked)
            }
        }
    }
```

**Comment by dca00** [comment]
Above are unacceptable solutions because having that much data being created and deleted every time the application starts increases the wear of storage media. This completely unnecessary bunch of 177 files in 67 folders and files has to have an option to be completely disabled.
All of that just for showing an HTML file to the user? Seriously?

**Reply by cremor** [reply]
You don't have to delete anything. Just set the path on startup and WebView2 will reuse the existing data folder.
This is all documented here: https://learn.microsoft.com/en-us/microsoft-edge/webview2/concepts/user-data-folder?tabs=dotnet
